akka-actor/src/main/scala/akka/io/dns/internal/AsyncDnsResolver.scala

```diff
@@ -0,0 +1,168 @@
+            case Some(r) ⇒
+              log.debug("Ipv6 cached {}", r)
+              Future.successful(r)
+            case None ⇒
+              sendQuestion(resolver, Question6(nextId(), caseFoldedName))
+          }
+        else
+          Empty
+
+        ipv4Recs.flatMap(ipv4Records ⇒ {
+          // TODO, do we want config to specify a max for this?
+          if (ipv4Records.nonEmpty) {
+            val minTtl4 = ipv4Records.minBy(_.ttl).ttl
+            cache.put((name, Ipv4Type), ipv4Records, minTtl4)
+          }
+          ipv6Recs.map(ipv6Records ⇒ {
+            if (ipv6Records.nonEmpty) {
+              val minTtl6 = ipv6Records.minBy(_.ttl).ttl
+              cache.put((name, Ipv6Type), ipv6Records, minTtl6)
+            }
+            ipv4Records ++ ipv6Records
+          })
+        })
+      case Srv ⇒
+        cache.get((name, Ipv4Type)) match {
+          case Some(r) ⇒ Future.successful(r)
+          case None ⇒
+            sendQuestion(resolver, SrvQuestion(nextId, caseFoldedName))
+        }
+
+    }
+    // FIXME, do we need this? it was in the old version, no test covered it.
+    /*
+    val canonicalNames: Map[String, String] = all.collect({
+      case CNameRecord(incomingName, _, cName) =>
+        incomingName.toLowerCase -> cName.toLowerCase
+    })(breakOut)
+
+    @tailrec
+    def resolveCanonicalName(name: String): String = {
+      canonicalNames.get(name) match {
+        case Some(cName) => resolveCanonicalName(cName)
+        case None => name
+      }
+    }
+
+    val canonicalName = resolveCanonicalName(name)
+    */
+
+    recs.map(result ⇒ DnsProtocol.Resolved(name, result))
+  }
+
+}
+
+/**
+ * INTERNAL API
+ */
+@InternalApi
+private[io] object AsyncDnsResolver {
+
+  private val ipv4Address =
+    """^[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}$""".r
+
+  private val ipv6Address =
+    """^\s*((([0-9A-Fa-f]{1,4}:){7}([0-9A-Fa-f]{1,4}|:))|(([0-9A-Fa-f]{1,4}:){6}(:[0-9A-Fa-f]{1,4}|((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3})|:))|(([0-9A-Fa-f]{1,4}:){5}(((:[0-9A-Fa-f]{1,4}){1,2})|:((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3})|:))|(([0-9A-Fa-f]{1,4}:){4}(((:[0-9A-Fa-f]{1,4}){1,3})|((:[0-9A-Fa-f]{1,4})?:((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3}))|:))|(([0-9A-Fa-f]{1,4}:){3}(((:[0-9A-Fa-f]{1,4}){1,4})|((:[0-9A-Fa-f]{1,4}){0,2}:((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3}))|:))|(([0-9A-Fa-f]{1,4}:){2}(((:[0-9A-Fa-f]{1,4}){1,5})|((:[0-9A-Fa-f]{1,4}){0,3}:((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3}))|:))|(([0-9A-Fa-f]{1,4}:){1}(((:[0-9A-Fa-f]{1,4}){1,6})|((:[0-9A-Fa-f]{1,4}){0,4}:((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3}))|:))|(:(((:[0-9A-Fa-f]{1,4}){1,7})|((:[0-9A-Fa-f]{1,4}){0,5}:((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3}))|:)))(%.+)?\s*$""".r
```

Comment:
Concise!

Reply:
I do not take credit for this ;)

Reply:
Lots of great work here. Perhaps it’d be nice to recognise any prior art here though.

Reply:
The license headers do so: https://github.com/akka/akka/pull/25214/files#diff-ff500718711617ce0d811010e3075540R3 Where specifically do you mean otherwise.

Reply:
Thanks Konrad. I didn’t notice that.